pyresample/bucket/__init__.py

```diff
@@ -209,27 +208,47 @@ def _call_pandas_groupby_statistics(self, scipy_method, data, fill_value=None, s
         # Calculate the min of the data falling to each bin
         out_size = self.target_area.size
 
-        # merge into one Dataframe
-        df = dd.concat([dd.from_dask_array(self.idxs), dd.from_dask_array(weights)],
-                       axis=1)
-        df.columns = ['x', 'values']
+        # create the output bins
+        bins = da.linspace(0, out_size-1, out_size).astype('int')
 
-        if scipy_method == 'min':
-            statistics = df.map_partitions(lambda part: part.groupby(
-                                           np.digitize(part.x,
-                                                       bins=np.linspace(0, out_size, out_size)
-                                                       )
-                                           )['values'].min())
+        # sort idxs and weights based on weights
+        if statistic_method == 'min':
+            order = np.argsort(weights)
+        elif statistic_method == 'max':
+            order = np.argsort(weights)[::-1]
 
-        elif scipy_method == 'max':
-            statistics = df.map_partitions(lambda part: part.groupby(
-                                           np.digitize(part.x,
-                                                       bins=np.linspace(0, out_size, out_size)
-                                                       )
-                                           )['values'].max())
+        @dask.delayed
+        def get_block_statistic(block, mask_buffer, idxs_sorted, weights_sorted):
+            extended_bins = np.append(block, np.iinfo(block.dtype).max)
+            # use the preallocated buffers
+            mask_buffer[:len(bins), :-1] = (bins[:, None] <= idxs_sorted[None, :]) &\
+                                           (extended_bins[1:, None] > idxs_sorted[None, :])
+            mask = mask_buffer[:len(bins), ...]
+            weight_idx = np.argmax(mask, axis=-1)
 
-        # fill missed index
-        statistics = (statistics + pd.Series(np.zeros(out_size))).fillna(0)
+            return weights_sorted[weight_idx]
+
+        idxs_sorted = self.idxs[order]
+        weights_sorted = np.append(weights[order], np.nan)
+
+        bins = np.linspace(0, out_size-1, out_size).astype('int')
+        chunk_size = len(bins)//int(1e3)
+        chunk_size = len(bins) if chunk_size == 0 else chunk_size
+
+        # preallocate buffers to avoid array creation in main loop
+        mask_buffer = np.empty((chunk_size, len(self.idxs) + 1), dtype=bool)
+        mask_buffer[:, -1] = True
+
+        # calculate the statistics in each block
+        blocks = da.from_array(bins, chunks=chunk_size).to_delayed()
```

Comment:
I don't think you need `.to_delayed` here. If it is a dask array then dask will figure out what it needs to do to pass it to a delayed function.

Reply:
Oh also you are iterating over `blocks` in the next block of code so you are creating a dask array and then immediately computing it so it can be iterated over, right? `for block in blocks`